### PR TITLE
Deploy front application from circleci

### PR DESCRIPTION
**Pull request by lunika**
### Purpose

We want to deploy frontend application on AWS. Files will be hosted on a S3 bucket and a cloudfront distribution is used to access them. 
To do that we created a terraform project describing all we need on AWS to deploy and host the front application. 
Then circleci is used to build and deploy the application every time a branch is merged in the main branch

### Propoal

- [x] create a terraform to manage the shared state
- [x] create a terraform config to manage resources on AWS
- [x] configure circleci to build and deploy front application

### CODE REVIEW

File: bin/build [hunk 1]
```diff
@@ -18,16 +19,23 @@ ARGUMENTS:
 #
 # Usage: build TAG
 function build() {
-    tag="${1}"
-    version=$(cat docker/images/backend/UPSTREAM_VERSION)
+    service="${1}"
+
+    if [[ "${service}" != "backend" ]] && [[ "${service}" != "frontend" ]]; then
+      echo -e "service ${service} must be backend or frontend"
```
Comment (madmatah) [inline bin/build]: Why `-e`  ?

File: docker/images/frontend/Dockerfile [hunk 2]
```diff
@@ -7,10 +7,14 @@ FROM busybox as downloader
...
-    rm frontend-master.tar.gz && \
-    mv frontend-master frontend
+# Moodlenet frontend version that you want to build (git commit, tag or branch name)
+# related to the official repository : https://gitlab.com/moodlenet/frontend/
+ARG MOODLENET_VERSION=master
+
+RUN wget https://gitlab.com/moodlenet/frontend/-/archive/${MOODLENET_VERSION}/frontend-${MOODLENET_VERSION}.tar.gz && \
+    tar xzf frontend-${MOODLENET_VERSION}.tar.gz && \
+    rm frontend-${MOODLENET_VERSION}.tar.gz && \
+    mv frontend-${MOODLENET_VERSION} frontend
```
Comment (madmatah) [inline docker/images/frontend/Dockerfile:17]: Suggestion : In the `builder` stage, you might declare `ENV` variables that are used during the build phase.

File: docker/images/frontend/UPSTREAM_VERSION [hunk 3]
```diff
@@ -0,0 +1 @@
+master
```
Comment (madmatah) [inline docker/images/frontend/UPSTREAM_VERSION]: Can you pin a specific version, in order to have consistent builds ?

**Review by madmatah — COMMENTED** [review]
LGTM, but another look by a terraform expert would be appreciated ;-) 

Found 2 typos in commit messages : 

1) in 39a32afba11c274bc466f7e7f1773224cf347c95 :
> To deploy frontend app we choos

-> choose

2) In 71c318f345e9f24af55ae0f75302d913f09768ec
> frontend application must be build

-> built

File: bin/build (hunk 1 above)
Comment (lunika) [reply bin/build]: copy past :-D

File: docker/images/frontend/Dockerfile (hunk 2 above)
Comment (lunika) [reply docker/images/frontend/Dockerfile:17]: You mean moving them from the top to the builder stage ?

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,6 +8,10 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
 
+### Added
+
+- deploy front application from circleci
```
Comment (sampaccoud) [inline CHANGELOG.md]:
```suggestion
- deploy front application to preprod from circleci
```

File: terraform/user.tf [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+  policy = <<POLICY
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Sid": "Cloudfront",
+      "Effect": "Allow",
+      "Action": "cloudfront:CreateInvalidation",
+      "Resource": "${aws_cloudfront_distribution.moodlenet_cloudfront_distribution.arn}"
```
Comment (sampaccoud) [inline terraform/user.tf:66]: Do you really need cache invalidation? You are not making a call to it in the CI configuration. Are your frontend files names unique by their content?

**Review by sampaccoud — APPROVED** [review]

**Pull request merged**